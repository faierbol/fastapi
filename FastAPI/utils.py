from datetime import datetime
from typing import Optional, TypedDict

import jwt
from fastapi import Depends
from FastAPI import config
from FastAPI.config import ALGORITHM, database
from fastapi.security import OAuth2PasswordBearer
from pydantic import EmailStr

optional_login = OAuth2PasswordBearer(tokenUrl="auth/login", auto_error=False)
oauth2_schema = OAuth2PasswordBearer(tokenUrl="auth/login")


class JWTPayload(TypedDict):
    user_id: int
    sub: EmailStr
    exp: datetime


def decode_jwt(token: str) -> JWTPayload:
    return jwt.decode(token, config.SECRET_KEY, algorithms=[ALGORITHM])  # type: ignore


async def get_user_id(token: str) -> Optional[int]:
    query = "SELECT user_id FROM authtoken_token WHERE key = :key"
    row = await database.fetch_one(query=query, values={"key": token})
    return row["user_id"] if row else None


async def get_current_user_or_none(
    token: str = Depends(optional_login),
) -> Optional[int]:
    if token:
        return decode_jwt(token)["user_id"]
    else:
        return None


async def get_current_user(token: str = Depends(oauth2_schema)) -> int:
    return decode_jwt(token)["user_id"]
