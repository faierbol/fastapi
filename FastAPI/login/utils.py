from datetime import datetime, timedelta

import jwt
from FastAPI.config import ACCESS_TOKEN_EXPIRE_MINUTES, ALGORITHM, SECRET_KEY
from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def create_access_token(subject: str, user_id: int) -> str:
    expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode = {"exp": expire, "sub": str(subject), "user_id": user_id}
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt


def verify_password(plain_password: str, hashed_password: str) -> bool:
    return pwd_context.verify(plain_password, hashed_password)  # type: ignore
