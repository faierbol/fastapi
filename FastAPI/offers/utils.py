from typing import Optional

from FastAPI.offers import crud
from FastAPI.offers.schemas import Status, ValidOffer
from pydantic import ValidationError


async def review_status_validator(offer_id: str) -> Optional[str]:
    offer = await crud.get_offer(offer_id) or {}
    try:
        ValidOffer(**offer)
        return None
    except ValidationError as e:
        return e.json()


async def set_review_status(offer_id: str) -> None:
    await crud.offer_set_status(offer_id=offer_id, status=Status.ACTIVE.value)
