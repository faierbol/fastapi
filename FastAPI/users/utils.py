import logging
import os.path
from datetime import date, datetime, timedelta
from typing import Any, Dict

import emails
import jwt
from emails.template import JinjaTemplate
from FastAPI import config
from FastAPI.config import ALGORITHM, SECRET_KEY
from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def get_password_hash(password: str) -> str:
    return pwd_context.hash(password)  # type: ignore


def get_activation_jwt(email: str) -> str:
    delta = timedelta(hours=config.EMAIL_RESET_TOKEN_EXPIRE_HOURS)
    now = datetime.utcnow()
    expires = now + delta
    exp = expires.timestamp()
    return jwt.encode(
        {
            "exp": exp,
            "nbf": now,
            "sub": email,
        },
        key=SECRET_KEY,
        algorithm=ALGORITHM,
    )


def send_email(
    email_to: str,
    subject_template: str,
    html_template: str,
    environment: Dict[str, Any],
) -> None:
    message = emails.Message(
        subject=JinjaTemplate(subject_template),
        html=JinjaTemplate(html_template),
        mail_from=(config.EMAILS_FROM_NAME, config.EMAILS_FROM_EMAIL),
    )
    smtp_options = {"host": config.SMTP_HOST, "port": config.SMTP_PORT}
    if config.SMTP_TLS:
        smtp_options["tls"] = True
    if config.SMTP_USER:
        smtp_options["user"] = config.SMTP_USER
    if config.SMTP_PASSWORD:
        smtp_options["password"] = config.SMTP_PASSWORD
    response = message.send(to=email_to, render=environment, smtp=smtp_options)
    logging.info(f"send email result: {response}")


def send_new_account_email(
    email_to: str,
    username: str,
    activation_token: str,
) -> None:
    subject = f"Phoqer - New account for user {username}"
    with open(os.path.join(os.getcwd(), "email-templates/build/new_account.html")) as f:
        template_str = f.read()
    link = f"http://phoqer.com/api/v2/users/activation/{activation_token}"
    send_email(
        email_to=email_to,
        subject_template=subject,
        html_template=template_str,
        environment={
            "project_name": "Phoqer",
            "username": username,
            "email": email_to,
            "link": link,
            "registration_date": date.today().strftime("%d-%m-%Y"),
        },
    )
