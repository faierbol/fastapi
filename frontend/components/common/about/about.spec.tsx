import { render } from '../../../tests/tests.utils';
import About from './index';

describe('About component', () => {
    it('matches snapshot', () => {
        const { asFragment } = render(<About />, {});
        expect(asFragment()).toMatchSnapshot();
    });
});
