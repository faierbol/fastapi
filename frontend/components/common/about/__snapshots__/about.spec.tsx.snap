// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrp-0-2-1"
  >
    <div
      class="slick-slider slider-0-2-2 slick-initialized"
      dir="ltr"
    >
      <button
        class="slick-arrow slick-prev"
        data-role="none"
        style="display: block;"
        type="button"
      >
         Previous
      </button>
      <div
        class="slick-list"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="-1"
            style="width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <picture
                class="img-0-2-3"
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              >
                <source
                  media="(max-width: 640px)"
                  srcset="/about/camera640.jpeg"
                />
                <source
                  media="(max-width: 1280px)"
                  srcset="/about/camera1280.jpeg"
                />
                <source
                  media="(max-width: 1280px)"
                  srcset="/about/camera1280.jpeg"
                />
                <source
                  media="(max-width: 1920px)"
                  srcset="/about/camera1920.jpeg"
                />
                <img
                  alt="phoqer"
                  class="img-0-2-3"
                  src="/about/camera.jpeg"
                />
              </picture>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="2"
            style="outline: none; width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="3"
            style="outline: none; width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="4"
            style="outline: none; width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="5"
            style="outline: none; width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="6"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <picture
                class="img-0-2-3"
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              >
                <source
                  media="(max-width: 640px)"
                  srcset="/about/camera640.jpeg"
                />
                <source
                  media="(max-width: 1280px)"
                  srcset="/about/camera1280.jpeg"
                />
                <source
                  media="(max-width: 1280px)"
                  srcset="/about/camera1280.jpeg"
                />
                <source
                  media="(max-width: 1920px)"
                  srcset="/about/camera1920.jpeg"
                />
                <img
                  alt="phoqer"
                  class="img-0-2-3"
                  src="/about/camera.jpeg"
                />
              </picture>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="7"
            style="width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="8"
            style="width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="9"
            style="width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="10"
            style="width: 0px;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="11"
            style="width: 0px;"
            tabindex="-1"
          />
        </div>
      </div>
      <button
        class="slick-arrow slick-next"
        data-role="none"
        style="display: block;"
        type="button"
      >
         Next
      </button>
    </div>
    <h2
      class="title-0-2-4"
    >
      phoqer
    </h2>
    <div
      class="content-0-2-5"
    >
      <p
        class="text-0-2-6"
      >
        Ты – это то, что тебя окружает. Мы полностью согласны с этой фразой. Вещи, которые мы покупаем, формируют человека.
      </p>
      <p
        class="text-0-2-6"
      >
        Поэтому мы собрали самые разные товары у нас на сайте, чтобы каждый смог найти желаемое. Легко. Быстро. По приемлемой цене и хорошего качества.
      </p>
      <p
        class="text-0-2-6"
      >
        Мы захотели создать что-то лучше, чем обычный супермаркет, где мы ходим с тележкой среди прилавков, в надежде найти конкретную вещь. Онлайн-шоппинг избавляет нас от многих неудобств. Здесь не придется стоять в очереди на кассе.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
